### PR TITLE
Minimal L1 Report Builder

### PR DESCRIPTION
**Pull request by NoahCarnahan**

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 19 errors:

* corehq/apps/hqwebapp/models.py, line 271 - E124 closing bracket does not match visual indentation
* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/reports/builder/__init__.py, line 28 - E122 continuation line missing indentation or outdented
* corehq/apps/userreports/reports/builder/forms.py, line 304 - E501 line too long (152 > 115 characters)
* corehq/apps/userreports/reports/builder/forms.py, line 337 - E121 continuation line under-indented for hanging indent
* corehq/apps/userreports/reports/builder/forms.py, line 376 - E501 line too long (116 > 115 characters)
* corehq/apps/userreports/reports/builder/forms.py, line 473 - E303 too many blank lines (2)
* corehq/apps/userreports/reports/builder/forms.py, line 550 - E501 line too long (120 > 115 characters)
* corehq/apps/userreports/tests/test_app_manager_integration.py, line 62 - E303 too many blank lines (2)
* corehq/apps/userreports/views.py, line 14 - F401 'View' imported but unused
* corehq/apps/userreports/views.py, line 16 - F401 'Application' imported but unused
* corehq/apps/userreports/views.py, line 17 - F401 'ParentCasePropertyBuilder' imported but unused
* corehq/apps/userreports/views.py, line 34 - F401 'DEFAULT_CASE_PROPERTY_DATATYPES' imported but unused
* corehq/apps/userreports/views.py, line 94 - E501 line too long (118 > 115 characters)
* corehq/apps/userreports/views.py, line 401 - E302 expected 2 blank lines, found 1
* corehq/apps/style/static/style/ko/knockout_bindings.ko.js, line 675 - ['value'] is better written in dot notation.
* corehq/apps/style/static/style/ko/knockout_bindings.ko.js, line 637 - Weird construction. Is 'new' necessary?

File: corehq/apps/userreports/reports/builder/forms.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+        ]
+
+        self.helper = FormHelper()
+        self.helper.form_class = "form-horizontal"
+        self.helper.form_id = "report-builder-form"
+        self.helper.layout = crispy.Layout(
+            crispy.Fieldset(
+                _('Create New Report'),
+                'report_type',
+                crispy.Field('application', data_bind='value: application'),
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: probably worth cross-linking this and the js with comments

File: corehq/apps/userreports/reports/builder/forms.py [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+                crispy.Field('source_type', data_bind='value: sourceType'),
+                crispy.Field('report_source', data_bind='''
+                    options: sourcesMap[application()][sourceType()],
+                    optionsText: function(item){return item.text},
+                    optionsValue: function(item){return item.value}
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: would be awesome to reuse this functionality in the "create data source from form/case" UI

File: corehq/apps/userreports/reports/builder/forms.py [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+            ds_builder = DataSourceBuilder(self.domain, app, source_type, report_source)
+
+            existing_sources = DataSourceConfiguration.by_domain(self.domain)
+            if len(existing_sources) >= 5:
+                suitable_source_exists = False
+                for s in existing_sources:
+                    if (
+                        s.referenced_doc_type == ds_builder.source_doc_type and
+                        s.configured_filter == ds_builder.filter and
+                        s.configured_indicators == ds_builder.indicators
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: probably easier to add explicit references to app/form/case here and timestamp them or link to a build id

File: corehq/apps/userreports/reports/builder/forms.py [hunk 4]
```diff
@@ -0,0 +1,569 @@
...
+        with slashes which requires this to be different from "id"
+        "text" will be used as the visible text in selects/select2s
+        "type" is "question", "case_property", or "meta"
+        For questions, "source" is the dict returned by Xform.get_questions, for
+        case properties and form metadata it is simply the name of the property.
+        """
+
+        def escape_id(id):
+            # TODO: This is fairly naive escaping
+            return id.strip("/").replace("/", "__")
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: can use something like this: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/sql.py#L80

File: corehq/apps/userreports/views.py [hunk 5]
```diff
@@ -80,6 +202,26 @@ def _edit_report_shared(request, domain, config):
 @require_POST
 def delete_report(request, domain, report_id):
     config = get_document_or_404(ReportConfiguration, domain, report_id)
+
+    # Delete the data source too if it's not being used by any other reports.
+    # If we decide to go this route, we most likely should write a new couch
+    # view to quickly get the ReportConfigurations that reference this data
+    # source.
```
Comment (snopoke) [inline corehq/apps/userreports/views.py]: New couch view seems like a good idea.

File: corehq/apps/userreports/tasks.py [hunk 6]
```diff
@@ -31,6 +31,8 @@ def rebuild_indicators(indicator_config_id):
         except DataError as e:
             logging.exception('problem saving document {} to table. {}'.format(doc['_id'], e))
     adapter.engine.dispose()
+    config.built = True
+    config.save()
```
Comment (snopoke) [inline corehq/apps/userreports/tasks.py]: Could probably use `django-soil` here to keep track of progress and completion status.

File: corehq/apps/userreports/reports/builder/forms.py [hunk 7]
```diff
@@ -0,0 +1,569 @@
...
+                       'submit',
+                       _(self.button_text)
+                   )
+               ),
+            ),
+        )
+
+    @property
+    def column_config_template(self):
+        path = os.path.join(
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: i think you can use the django template loader for this

**Comment by snopoke** [comment]
Did a quick pass and looks good. Do we have an idea of how to handle app changes when they happen?
Does the user just come back and update the report config?

File: corehq/apps/userreports/views.py (hunk 5 above)
Comment (czue) [inline corehq/apps/userreports/views.py]: +1

File: corehq/apps/userreports/reports/builder/forms.py [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+            # The uuid gets truncated, so it's not really universally unique.
+            table_id=_clean_table_name(self.domain, str(uuid.uuid4().hex)),
+            configured_filter=self.ds_builder.filter,
+            configured_indicators=self.ds_builder.indicators
+        )
+        data_source_config.validate()
+
+        # Check if a suitable data source already exists.
+        # This checking is pretty naive. Do something better.
+        # Also inefficient. Consider writing a new couch view.
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: +1 - same comment as above

**Comment by NoahCarnahan** [comment]
Yeah that's right @snopoke, if you make any app changes that affect the report you need to edit the config "by hand" or use the report build to create the report again. (Basically we're just push that problem out of scope for now :) )

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/views.py, line 14 - F401 'View' imported but unused
* corehq/apps/style/static/style/ko/knockout_bindings.ko.js, line 637 - Weird construction. Is 'new' necessary?

File: corehq/apps/reports/urls.py [hunk 9]
```diff
@@ -19,6 +25,10 @@
 urlpatterns = patterns('corehq.apps.reports.views',
     TestReport.url_pattern(),
     ConfigurableReport.url_pattern(),
+    url(r'^builder/create/$', CreateNewReportBuilderView.as_view(), name="create_new_report_builder"),
+    url(r'^builder/configure/bar/$', ConfigureBarChartReportBuilderView.as_view(), name="configure_bar_chart_report_builder"),
```
Comment (dimagimon) [inline corehq/apps/reports/urls.py]: E501 line too long (126 > 115 characters)

File: corehq/apps/reports/urls.py [hunk 10]
```diff
@@ -19,6 +25,10 @@
...
+    url(r'^builder/configure/pie/$', ConfigurePieChartReportBuilderView.as_view(), name="configure_pie_chart_report_builder"),
```
Comment (dimagimon) [inline corehq/apps/reports/urls.py]: E501 line too long (126 > 115 characters)

File: corehq/apps/reports/urls.py [hunk 11]
```diff
@@ -19,6 +25,10 @@
...
+    url(r'^builder/configure/table/$', ConfigureTableReportBuilderView.as_view(), name="configure_table_report_builder"),
```
Comment (dimagimon) [inline corehq/apps/reports/urls.py]: E501 line too long (121 > 115 characters)

File: corehq/apps/style/static/style/ko/knockout_bindings.ko.js [hunk 12]
```diff
@@ -629,6 +629,53 @@ ko.bindingHandlers.typeahead = {
     }
 };
 
+/**
+ * Converts the bound element to a select2 widget. The value of the binding is
+ * a list of strings, or a list of objects with the keys 'id' and 'text' used
+ * for the select2's options.
+ */
+ko.bindingHandlers.select2 = new function(){
```
Comment (dimagimon) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: Weird construction. Is 'new' necessary?

Comment (czue) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: is @dimagimon becoming sentient?

Comment (snopoke) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: Was also a bit puzzled by this.

Comment (NoahCarnahan) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: I wanted to create an object where both the `init` and `update` properties refer to the `SOURCE_KEY` property, so I couldn't use an object literal. I think the anonymous constructor approach that I used looks a little cleaner than either breaking the object's definition over multiple statements or using a "self-calling" anonymous function. Is there a best practice or style guideline for this situation?
(I found a somewhat related discussion on [SO](http://stackoverflow.com/questions/1536127/module-pattern-vs-instance-of-an-anonymous-constructor))

**Comment by dimagimon** [comment]
There are 1 errors:

* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file

**Comment by dimagimon** [comment]
There are 2 errors:

* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

File: corehq/apps/userreports/tasks.py (hunk 6 above)
Comment (NoahCarnahan) [inline corehq/apps/userreports/tasks.py]: I didn't know about `django-soil`, thanks for pointing that out.
As currently written, I check if the data source has been built every time the report is
loaded. That means I'll probably be checking the built status after the `soil.DownloadBase`
is no longer in the cache. Does soil have a good way to handle this situation?

I could still store the built status on the `DataSourceConfiguration`. Then, when the report
is loaded I would first look to get the status from the `DownloadBase`, and if it doesn't
exist, look on the `DataSourceConfiguration` itself. Seems sorta redundant though...

Comment (snopoke) [inline corehq/apps/userreports/tasks.py]: I was thinking more of being able to give more of a progress dialog to the user. You could still use this to determine if the config has been built but I would suggest setting `built = False` when changes get made to the config.

Comment (NoahCarnahan) [inline corehq/apps/userreports/tasks.py]: Makes sense, thanks.

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

File: corehq/apps/userreports/reports/builder/forms.py (hunk 1 above)
Comment (NoahCarnahan) [inline corehq/apps/userreports/reports/builder/forms.py]: Done. Thanks.

File: corehq/apps/userreports/reports/builder/forms.py (hunk 2 above)
Comment (NoahCarnahan) [inline corehq/apps/userreports/reports/builder/forms.py]: Added in 84bee6d. Unfortunately I didn't come up with a great way to encapsulate this functionality.

File: corehq/apps/userreports/reports/builder/forms.py (hunk 3 above)
Comment (NoahCarnahan) [inline corehq/apps/userreports/reports/builder/forms.py]: Updated in 5298395 to use app + xmlns/case type + app build number. There's a potentially big downside here though in that any app change will "invalidate" the data source, even if it doesn't affect the case type or form that the data source is based on. But, perhaps this is satisfactory until a more comprehensive solution is found (such as calculating diffs between needed data sources and existing data sources or something like that).

File: corehq/apps/userreports/reports/builder/forms.py (hunk 4 above)
Comment (NoahCarnahan) [inline corehq/apps/userreports/reports/builder/forms.py]: Fixed in 2b1fef6, thanks!

File: corehq/apps/userreports/views.py (hunk 5 above)
Comment (NoahCarnahan) [inline corehq/apps/userreports/views.py]: Fixed with b2cd306, thanks.

File: corehq/apps/userreports/reports/builder/forms.py (hunk 7 above)
Comment (NoahCarnahan) [inline corehq/apps/userreports/reports/builder/forms.py]: Implemented in cc1f5f5, thanks for the tip.

**Comment by NoahCarnahan** [comment]
Thanks for the feedback guys. I've the addressed the comments in the latest commits. Unfortunately it looks like I broke the non report builder "Edit data source" form though, so that needs to be fixed prior to merging. Otherwise, I think this is ready and I believe the plan is to not go through QA prior to merging since this is under a flag. Amelia has also suggested some UI changes that I'll be working on in ~2 weeks when my interrupt duty is over.

**Comment by NoahCarnahan** [comment]
@czue @snopoke 
I tracked down the bug I was chasing, but I'm not sure what the cleanest way to fix it is.
This is what happened:
- Some bad commits were introduced on master. It was one of these:
    029d2ea957a837c11801b0c4a10bf6d32fbb8f98
    be782c5774d6705a529cf67762b307bfe0efb0da
    bfd9392d0d552e98d5e7c4f5ab50cdf7fdedc71a
- I merged those commits into my branch when I merged master into my branch a while ago.
- At a later point, the bug was fixed by 8e3073e6ba001aa85c9e36b86aa401fcc588eeeb
- The bug is fixed on master, but if I merge master into my branch the bug still remains on my branch.

I'm not sure what the cleanest way to fix this is. I tried cherry-picking the commit that fixes it, but that does nothing. Should I rebase my branch off of master? I generally try to avoid rewriting history though. I suppose I could manually make the edits introduced by the fixing commit onto my branch.
Thoughts? I'm not sure why merging master in the first place didn't fix things.

(For reference, the bug is that saving changes on the data source configuration form has no effect - form is interpreted as invalid)

**Comment by czue** [comment]
@NoahCarnahan can you just read https://github.com/dimagi/commcare-hq/commit/8e3073e6ba001aa85c9e36b86aa401fcc588eeeb and figure out if there was a bad merge that caused not the whole fix to be applied?

The main issue was that the form arguments being passed in and up the inheritance chain were not consistent with what django's form classes expected so somewhere we were passing the wrong value for some arguments.

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

**Comment by NoahCarnahan** [comment]
@czue 
manually creating a commit that replicates 8e3073e did the trick.
In the course of debugging this I found a different bug though (form questions with no label don't show up in select2), so let me fix that before we merge.

**Comment by NoahCarnahan** [comment]
(False alarm on the other bug)

**Comment by czue** [comment]
@NoahCarnahan are you ready for "final" review now?

**Comment by NoahCarnahan** [comment]
Yup, I think so.

**Comment by czue** [comment]
@NoahCarnahan a few small comments, but I'm pretty much signed off. where did we land on putting this into a subscription? the only reason to do it now is so that we don't forget that we enabled it for domains that weren't on the right subscription now (which I think isn't a bad reason). otherwise when we do add that restriction we'll break them, right?

**Comment by dimagimon** [comment]
There are 7 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/models.py, line 208 - E501 line too long (116 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

**Comment by NoahCarnahan** [comment]
@czue I forgot about the subscription question. I wasn't really involved with that discussion, but @amsagoff might know where that stands? (feel free to take to email if more appropriate). Right now this is under the general "User Configurable Reports UI" flag. I think it's probably best to implement the subscription constraint now.

**Comment by czue** [comment]
@NoahCarnahan i think for now we can make it pro or higher. you can refer to @nickpell's recent PR for a template of how to do it: https://github.com/dimagi/commcare-hq/pull/5750

**Comment by NoahCarnahan** [comment]
Great, will do.

File: corehq/apps/userreports/reports/builder/forms.py (hunk 2 above)
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: thoughts on this attempt? https://github.com/dimagi/commcare-hq/pull/5793

**Comment by czue** [comment]
also i think the data source lookup by domain might not be filtering properly. i see all data sources in the sidebar on every domain

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/models.py, line 208 - E501 line too long (116 > 115 characters)
* corehq/apps/userreports/sql.py, line 96 - W391 blank line at end of file
* corehq/apps/userreports/views.py, line 343 - E303 too many blank lines (3)
* corehq/apps/style/static/style/ko/knockout_bindings.ko.js, line 637 - Weird construction. Is 'new' necessary?
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

**Comment by NoahCarnahan** [comment]
@czue, just fixed the source lookup by domain. I had forgotten the `end_key`.

**Comment by dimagimon** [comment]
There are 7 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 103 - W391 blank line at end of file
* corehq/apps/userreports/ui/forms.py, line 11 - F401 'memoized' imported but unused
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

**Comment by dimagimon** [comment]
There are 7 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 103 - W391 blank line at end of file
* corehq/apps/userreports/ui/forms.py, line 11 - F401 'memoized' imported but unused
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

**Comment by NoahCarnahan** [comment]
@czue I've put the report builder under the pro plan (see  b990e9f). It requires the running of a management command `./manage.py cchq_prbac_bootstrap`. Is it better with regard to our deployment process if I put this in a migration like Nick did here: a0057a9425abdc77ef906341038d601fc5edac65 ?

**Comment by nickpell** [comment]
+1 automating deploys

**Comment by dimagimon** [comment]
There are 7 errors:

* corehq/apps/reports/urls.py, line 29 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 30 - E501 line too long (126 > 115 characters)
* corehq/apps/reports/urls.py, line 31 - E501 line too long (121 > 115 characters)
* corehq/apps/userreports/sql.py, line 103 - W391 blank line at end of file
* corehq/apps/userreports/ui/forms.py, line 11 - F401 'memoized' imported but unused
* corehq/apps/userreports/_design/views/data_sources_by_build_info/map.js, line 1 - Missing name in function declaration.
* corehq/apps/userreports/_design/views/report_configs_by_data_source/map.js, line 1 - Missing name in function declaration.

**Comment by NoahCarnahan** [comment]
management command will now run in a migration.

**Comment by NoahCarnahan** [comment]
I think this beast is ready for merge.

**Comment by czue** [comment]
:eyes:

**Pull request merged**